# **Credit Card Customer Churn Analysis (ETL Pipeline)**

## Objectives (ETL)

In this notebook I will be running through the entire ETL pipeline. The goal here is to process the raw dataset so it is ready for the next stages: EDA(exploratory data analysis) & so i can import the processed dataset into PowerBi to cree=ate interactive dashboards which will be presented to both technical and non-technical stakeholders.  

## Hypotheses

* **H1:** Customers with shorter tenure are more likely to churn compared to customers with longer tenure.
* **H2:** Customers with lower credit utilisation ratios are more likely to churn than customers with higher utilisation ratios.
* **H3:** Churn rates differ significantly across income categories.




## Inputs

The inputs required to run this notebook are:

* A CSV file containing Credit Card Customer Churn dataset sourced from Kaggle.
* Python libraries including but not limited to Scikit-Learn, pandas, Numpy, matplotlib, and seaborn.

## Outputs
The outputs generated by this notebook include:

* Cleaned and processed DataFrame(s)
* Feature Engineered columns needed for hypothesis testing

---

# Section 1: Data Loading and Initial Exploration (Extract)

In this section, I will load the raw credit card churn dataset and perform an initial exploration to understand the structure, data types, and basic characteristics of the data.

In [1]:
# Import the pandas library for data manipulation
import pandas as pd 

# Load the dataset into a DataFrame
df = pd.read_csv("../data/raw_data/BankChurners.csv") 

# Check info and missing values
df.info()
df.isna().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

### The above output shows:

* No missing values — all 10,127 rows are complete.
* All columns have data — no nulls.
* Data types look correct for now: integers, floats, objects (categorical).

---

# Section 2: Feature engineering (Transform) & saving transformed dataset into a proccessed fike (Load)

Section 2 content

In [2]:
# Drop unnecessary columns
df = df.drop(columns=[
    'CLIENTNUM',
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1',
    'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2'
])


In [3]:
# Encode target variable
df['Attrition_Flag'] = df['Attrition_Flag'].map({'Existing Customer': 0, 'Attrited Customer': 1})


In [4]:
# Feature engineering
df['Tenure_Months'] = df['Months_on_book']
df['Avg_Trans_per_Month'] = df['Total_Trans_Amt'] / df['Months_on_book']
df['Credit_Utilization'] = df['Total_Trans_Amt'] / df['Credit_Limit']
df = df.drop(columns=['Months_on_book'])


In [5]:
# Encode categorical features
categorical_cols = ['Gender', 'Education_Level', 'Marital_Status', 'Income_Category', 'Card_Category']
df = pd.get_dummies(df, columns=categorical_cols, drop_first=True)


In [6]:
# Check processed dataset
df.head()


Attrition_Flag  Customer_Age  Dependent_count  Total_Relationship_Count  \
0               0            45                3                         5   
1               0            49                5                         6   
2               0            51                3                         4   
3               0            40                4                         3   
4               0            40                3                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0                       1                      3       12691.0   
1                       1                      2        8256.0   
2                       1                      0        3418.0   
3                       4                      1        3313.0   
4                       1                      0        4716.0   

   Total_Revolving_Bal  Avg_Open_To_Buy  Total_Amt_Chng_Q4_Q1  ...  \
0                  777          11914.0                 1.335  ...   
1                  864           7392.0                 1.541  ...   
2                    0           3418.0                 2.594  ...   
3                 2517            796.0                 1.405  ...   
4                    0           4716.0                 2.175  ...   

   Marital_Status_Single  Marital_Status_Unknown  Income_Category_$40K - $60K  \
0                  False                   False                        False   
1                   True                   False                        False   
2                  False                   False                        False   
3                  False                    True                        False   
4                  False                   False                        False   

   Income_Category_$60K - $80K  Income_Category_$80K - $120K  \
0                         True                         False   
1                        False                         False   
2                        False                          True   
3                        False                         False   
4                         True                         False   

   Income_Category_Less than $40K  Income_Category_Unknown  \
0                           False                    False   
1                            True                    False   
2                           False                    False   
3                            True                    False   
4                           False                    False   

   Card_Category_Gold  Card_Category_Platinum  Card_Category_Silver  
0               False                   False                 False  
1               False                   False                 False  
2               False                   False                 False  
3               False                   False                 False  
4               False                   False                 False  

[5 rows x 35 columns]

---

In [7]:
# Save processed dataset
df.to_csv("../data/cleaned_data/credit_churn_processed.csv", index=False)


---

## ETL Process Overview

* Import Libraries: Loaded pandas for data manipulation.
* Load Dataset: Imported raw CSV into a DataFrame.
* Inspect Dataset: Checked column types and missing values.
* Drop Columns: Removed IDs and precomputed Naive Bayes columns.
* Encode Target: Converted Attrition_Flag to numeric (0 = active, 1 = churned).
* Feature Engineering: Added Tenure_Months, Avg_Trans_per_Month, and Credit_Utilization.
* Encode Categorical: Applied one-hot encoding for categorical features.
* Save Processed Data: Exported cleaned dataset for EDA, ML, and dashboard.